kedro/io/catalog_config_resolver.py

```diff
@@ -0,0 +1,261 @@
+        """Calculate the specificity of a pattern based on characters outside curly brackets."""
+        # Remove all the placeholders from the pattern and count the number of remaining chars
+        result = re.sub(r"\{.*?\}", "", pattern)
+        return len(result)
+
+    @classmethod
+    def _sort_patterns(cls, dataset_patterns: Patterns) -> Patterns:
+        """Sort a dictionary of dataset patterns according to parsing rules.
+
+        In order:
+        1. Decreasing specificity (number of characters outside the curly brackets)
+        2. Decreasing number of placeholders (number of curly bracket pairs)
+        3. Alphabetically
+        """
+        sorted_keys = sorted(
+            dataset_patterns,
+            key=lambda pattern: (
+                -(cls._pattern_specificity(pattern)),
+                -pattern.count("{"),
+                pattern,
+            ),
+        )
+        catch_all = [
+            pattern for pattern in sorted_keys if cls._pattern_specificity(pattern) == 0
+        ]
+        if len(catch_all) > 1:
+            raise ValueError(
+                f"Multiple catch-all patterns found in the catalog: {', '.join(catch_all)}. Only one catch-all pattern is allowed, remove the extras."
+            )
+        return {key: dataset_patterns[key] for key in sorted_keys}
+
+    def list_patterns(self) -> list[str]:
+        """List al patterns available in the catalog."""
+        return (
+            list(self._dataset_patterns.keys())
+            + list(self._default_pattern.keys())
+            + list(self._runtime_patterns.keys())
+        )
+
+    def match_pattern(self, ds_name: str) -> str | None:
+        """Match a dataset name against patterns in a dictionary."""
+        all_patterns = self.list_patterns()
+        matches = (pattern for pattern in all_patterns if parse(pattern, ds_name))
+        return next(matches, None)
+
+    @classmethod
+    def _extract_patterns(
+        cls,
+        config: dict[str, dict[str, Any]] | None,
+        credentials: dict[str, dict[str, Any]] | None,
+    ) -> tuple[Patterns, Patterns]:
+        """Extract and sort patterns from the configuration."""
+        config = copy.deepcopy(config) or {}
+        credentials = copy.deepcopy(credentials) or {}
+        dataset_patterns = {}
+        user_default = {}
+
+        for ds_name, ds_config in config.items():
+            if cls.is_pattern(ds_name):
+                dataset_patterns[ds_name] = _resolve_credentials(ds_config, credentials)
+
+        sorted_patterns = cls._sort_patterns(dataset_patterns)
+        if sorted_patterns:
+            # If the last pattern is a catch-all pattern, pop it and set it as the default
+            if cls._pattern_specificity(list(sorted_patterns.keys())[-1]) == 0:
+                last_pattern = sorted_patterns.popitem()
+                user_default = {last_pattern[0]: last_pattern[1]}
+
+        return sorted_patterns, user_default
+
+    def _init_configs(
```

Comment:
`init_configs` isn't the most descriptive name IMO. Maybe `_resolve_config_credentials`?